### ISSUE TITLE
Incognito & Container Support

### ISSUE THREAD

**Issue by erosman** [foxyproxy/browser-extension#33]
# Incognito & Container Support

Support can be added for Firefox only.
See also: #27

## User Interface
Incognito can be supported without any issues. However to support containers as well. there are 2 options for the User Interface:

### 1. Generic User Interface
Use a generic container list, not knowing or matching the number of containers the user has e.g.
- Incognito
- Container 1
- Container 2
- Container 3
- Container 4

### 2. Specific User Interface
Get the container list [contextualIdentities.query()](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/contextualIdentities/query) and create the UI
   - [contextualIdentities](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/contextualIdentities)
     > To use this API you need to include the "contextualIdentities" and "cookies" [permissions](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions) in your [manifest.json](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json) file.
   - Adding  "contextualIdentities" permission automatically enables containers in Firefox which is undesirable to users who don't want to use containers
   - "contextualIdentities" permission is not available as an [optional_permissions](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/optional_permissions)
   - See also: [Bug 1386673: Make Contextual Identity extensions be an optional permission](https://bugzilla.mozilla.org/show_bug.cgi?id=1386673)

## Implementation
- Incognito/Containers can be set to a specific proxy (_not patterns or PAC URL_)
- Incognito/Container settings will have precedence over other settings (_except Tab Proxy_)

The checking order would be:

1. Tab Proxy
2. Incognito/Container Proxy
3. Check for "Disable" mode
4. Global Exclude
5. PAC URL
6. Proxy by Patterns or Individual Proxy

### To decide
- Should "Disable" disable everything?
- Should Tab Proxy be independent of "Disable" since it has its own independent on/off?
- Should Incognito/Container Proxy be independent of "Disable"?

## Sync and Preferences: Import/Export

Container settings may mismatch, if the user has different containers on different browsers.

**Comment by erosman** [comment]
FoxyProxy v8.1 (on repo) includes incognito & Container support for Firefox only.
The Log is also updated to include incognito & Container details.

The feature is still under consideration and development.

**Comment by pmarks-net** [comment]
The Incognito feature doesn't seem to do anything on Firefox 118.  Here's what I tried:

- Copy manifest-firefox.json to manifest.json
- Load manifest.json from about:debugging
- Run in Private Windows > Allow
- Proxies > add 10.2.2.1, SOCKS5, 1080
- Options > Incognito > 10.2.2.1:1080
![incognito](https://github.com/foxyproxy/browser-extension/assets/4578188/53691cf1-1f7d-4305-9f15-de7f184c85c7)
- Open a private window
- https://ifconfig.me/ still shows non-proxied IP address
- Click FoxyProxy icon > 10.2.2.1:1080
- https://ifconfig.me/ shows proxied IP address (but this affects regular tabs too)

**Comment by erosman** [comment]
I have uploaded the under-development v8.1 for another issue (#35). Therefore, the incognito & container is not finalised yet.

What does the log show?

**Comment by TriMoon** [comment]
wrt. "Tab Proxy":

It would be nice to have an option "DIRECT" in the drop-down list to select as proxy to use for the tab, which would temporarily disable all proxies for the tab :wink: 
- A workaround could be to create such a proxy manually and select it, but it's nicer to have that functionality in your UI by default :smile_cat: 
- Even when you select a tab proxy in the drop-down list and click and "Set Tab Proxy" it is not shown when you re-open the popup or reload the page, which might be a good idea also. :+1:

**Comment by erosman** [comment]
> It would be nice to have an option "DIRECT" (or "Disabled") in the drop-down list to select as proxy to use for the tab, which would temporarily disable all proxies for the tab 
There are set & unset to enable and disable Tab proxy.

I can look into adding a DIRECT option to Set Tab Proxy which would disable all proxying related to that tab.

> Even when you select a tab proxy in the drop-down list and click and "Set Tab Proxy" it is not shown when you re-open the popup or reload the page, which might be a good idea also.

Set Tab Proxy is an independent function and its data is not saved. The page-action icon is set to show if a tab has Tab Proxy or not. It is possible to grab the data but that would involve a number of async processes which have to be performed every time the popup is shown but not applicable 99% of the times.

**Comment by pmarks-net** [comment]
> It is possible to grab the data but that would involve a number of async processes which have to be performed every time the popup is shown but not applicable 99% of the times.

As the developer, you're free to do whatever you want, but if your code isn't doing the hard work to keep track of its own state, then what hope do the users have?

**Comment by erosman** [comment]
> As the developer, you're free to do whatever you want, but if your code isn't doing the hard work to keep track of its own state, then what hope do the users have of understanding what's happening?

If there is popular demand for a feature, it will be looked into.

> I notice that the pageAction icon disappears when I reload the page, for example.

That shouldn't happen. I will look into it.

**Comment by pmarks-net** [comment]
> If there is popular demand for a feature, it will be looked into.

I'm talking more about a design philosophy.  When adding a feature, often the hardest part is making it comprehensible to the user: https://en.wikipedia.org/wiki/The_Design_of_Everyday_Things

**Comment by erosman** [comment]
The performance also matters. 

> [Add-on Policies: Development Practices](https://extensionworkshop.com/documentation/publish/add-on-policies/)
> Add-ons must not negatively impact the performance or stability of Firefox.

**Comment by ericjung** [comment]
> Set Tab Proxy" it is not shown when you re-open the popup or reload the page, which might be a good idea also.

Is my understanding correct that FP is not displaying the proxy in use when the popup is opened? And doing so would be a significant performance hit? If so, we should probably remove the Set Tab Proxy feature altogether. I hope I misunderstand.

**Comment by erosman** [comment]
- Tab Proxy is shown by an icon in the address-bar
- Checking & showing it also in the toolbar popup adds extra processes every time the popup is opened, and would be considered if there is a popular demand for it


Please note that **Tab Proxy** is a separate feature than **Incognito & Container Proxy** and should be discussed in another topic.

**Comment by TriMoon** [comment]
@erosman excuse me but your argument is flawed wrt to performance penalty for showing the TabProxy in the popup when it is opened, the flaw is that your code would only do the extra work when the popup is opened by the user which doesn't happen that often anyway during normal browsing...
But WHEN the user opens the popup it expects to see ALL relevant info that is applied at that moment...

**Comment by erosman** [comment]
Please limit the discussion to **Incognito & Container Support** in this topic.

**Comment by ntninja** [comment]
Feedback about this new option:

  1. It looks like a promising way of consolidating the current “Container Proxy” + “FoxyProxy Standard” setup into just using FoxyProxy!
  2. The fact that the UI just shows Container 1 – 4 instead of names really looks like a bug at first glance and the reasoning behind should be explained then and there
  3. Unfortunately, we have a setup with more than 4 containers in total and one of the containers that needs to be set up to use a proxy happens to not be configurable using the current UI so using just FoxyProxy is infeasible
      * Assuming FoxyProxy won’t somehow gain the ability to find out the total number of containers maybe some kind of “Add another” link to make the list longer would be very helpful without changing too much
  4. If two or more configured proxies share the same hostname and port configuration, selecting any of them as the proxy to use for a container will always result in the first proxy with that hostname/port combination in the list of proxies being applied on “Save” (bug)
    * This occurs specifically because the configuration contains several entries for the local TOR node (localhost:9050) that only differ on the username and password set: Using TOR this way causes the TOR node establish independent network paths (“stream isolation”) for each of them even when visiting the same page from different contexts in the hope that this will make it harder/impossible to detect that these connections originate from the same host. (The sent username/password combination is just used as an identifier for the stream isolation context and has nothing to do with authentication.)

**Comment by erosman** [comment]
1. Incognito/Container feature is experimental
2. It is explained in the help
3. It is not possible to get any details about the user's containers without the right permission. The “Add" is a possibility that can be looked into.
4. FoxyProxy regularly searches the data to [find()](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array/find) the right proxy using `hostname:port` & PAC URL as the criteria. While having multiple proxies with the same `hostname:port` are permitted, the search would always return the first one. 
The request requires further investigation & discussion.

**Comment by ntninja** [comment]
1. I didn’t know that until you pointed it out (it’s not obvious at all) – just adding the beaker-icon from the help to the main view would change that though
2. My first instinct at least is not to consult the help when I’m like “why is this broken?”, doubt most other people feel very differently about this – maybe an explicit pointer to the help then?
    (ie: *Incognito/Container proxy (Firefox only, see help for known issues and limitations)*)
3. Understandable! Maybe also link to the Mozilla bug in the help to make it clear that that is what needs to happen *to be able* to improve the situation?
4. That explains why this happens from a technical implementation PoV, but it doesn’t mean that it should be this way: It took me quite a bit of time to even figure out that it’s the duplicate username+port combination that triggers this behaviour because the effective UX is “it just changes to something else on save” (with no further explanation) and that is clearly suboptimal.

I get that this should be considered an experimental PoC for now though – just wanted to give some feedback in the hopes that future iterations of this may be improved! 🙂

**Comment by erosman** [comment]
> Understandable! Maybe also link to the Mozilla bug in the help to make it clear that that is what needs to happen to be able to improve the situation?

It is how the "contextualIdentities" API is designed.

> [contextualIdentities](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/contextualIdentities)
> 
> To use this API you need to include the "contextualIdentities" and "cookies" permissions in your `manifest.json` file.

Unfortunately, it is not available as [optional_permissions](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/optional_permissions)

- [Bug 1386673: Make Contextual Identity extensions be an optional permission](https://bugzilla.mozilla.org/show_bug.cgi?id=1386673)

**Comment by erosman** [comment]
@ntninja Repo is updated with v8.8 and a workaround for issue no 4.

Please read [Help: Proxies](https://foxyproxy.github.io/browser-extension/src/content/help.html#proxies) for details.

Let ms know if it works out.

**Comment by erosman** [comment]
@ntninja Repo is updated with v8.8 and a workaround for issue no 4.

Please read [Help: Proxies](https://foxyproxy.github.io/browser-extension/src/content/help.html#proxies) for details.

Let me know how it works out.

**Comment by ntninja** [comment]
@erosman: I had the workaround to use `<somename>.localhost` for each of those since they all resolve `127.0.0.1` internally but are different hostnames, so in *our* case another workaround wasn’t necessary – for #76 it wasn’t an option though, I do realize that. I’ll try your workaround once it lands though.

**Comment by erosman** [comment]
> I had the workaround to use `<somename>.localhost` for each of those since they all resolve `127.0.0.1` internally but are different hostnames, so in our case another workaround wasn’t necessary

@ntninja TBH, using `<somename>.localhost` is more elegant. :+1:

**Comment by debilin** [comment]
The Firefox extension [cookie-autodelete](https://github.com/Cookie-AutoDelete/Cookie-AutoDelete) was able to grab information about containers. I don't know how they do it, but we could look at their code and attempt to replicate.

**Comment by erosman** [comment]
> The Firefox extension [cookie-autodelete](https://github.com/Cookie-AutoDelete/Cookie-AutoDelete) was able to grab information about containers. I don't know how they do it, but we could look at their code and attempt to replicate.

Getting container details is not a problem. Above extension has "contextualIdentities" permission in its [manifest.json](https://github.com/Cookie-AutoDelete/Cookie-AutoDelete/blob/ca628ccfa567eeb5e088c74812090659185132b2/extension/manifest.json#L44C6-L44C26).

As explained under "2. Specific User Interface", adding "contextualIdentities" permission is fine for people who use containers.
However, for people who don't want to use container, adding "contextualIdentities" permission will automatically enables containers which is undesirable.

At the moment, there is no other option.

**Comment by equeim** [comment]
Doesn't seem to work for me in Firefox 123 with FoxyProxy 8.9. Proxy is applied for incognito window, but not for containers.

**Comment by erosman** [comment]
> Doesn't seem to work for me in Firefox 123 with FoxyProxy 8.9. Proxy is applied for incognito window, but not for containers.

Are you using containers (enabled via other extensions) and have you set them to a proxy in FoxyProxy Options?

**Comment by equeim** [comment]
> > Doesn't seem to work for me in Firefox 123 with FoxyProxy 8.9. Proxy is applied for incognito window, but not for containers.
> 
> Are you using containers (enabled via other extensions) and have you set them to a proxy in FoxyProxy Options?

Yes

**Comment by erosman** [comment]
Can post your settings (remove/obfuscate any personal data)?

**Comment by equeim** [comment]
Sure:
```json
{
  "mode": "disable",
  "sync": false,
  "autoBackup": false,
  "passthrough": "",
  "theme": "",
  "container": {
    "incognito": "",
    "container-1": "127.0.0.1:9050",
    "container-2": "",
    "container-3": "",
    "container-4": ""
  },
  "commands": {
    "setProxy": "",
    "setTabProxy": "",
    "quickAdd": ""
  },
  "data": [
    {
      "active": true,
      "title": "TOR",
      "type": "socks5",
      "hostname": "127.0.0.1",
      "port": "9050",
      "username": "",
      "password": "",
      "cc": "",
      "city": "",
      "color": "#8b4513",
      "pac": "",
      "pacString": "",
      "proxyDNS": true,
      "include": [],
      "exclude": []
    }
  ]
}
```

Tabs in the first container do not use the proxy.

**Comment by erosman** [comment]
@equeim What does the log show?

**Comment by equeim** [comment]
The log shows that url in container tab is opened without proxy.

I created a new Firefox profile and it works there. I then tried to disable all extensions except FoxyProxy in my original profile and it still doesn't work. IDK what breaks it. I'm using "Firefox Multi-Account Containers" and it's native proxy feature for now.

**Comment by erosman** [comment]
> I created a new Firefox profile and it works there.

If the feature works in a new profile without any other extensions, then the issue is a conflict with another extension that handles containers e.g. "Firefox Multi-Account Containers".
Furthermore, "Firefox Multi-Account Containers" also has a proxy feature which can conflict with other proxying processes.

If that is the case, there isn't much that can be done.

The only suggestion at the moment is to disable "Firefox Multi-Account Containers" and enable containers by another method (i.e. use Firefox Nightly, Dev version, or another extension which enables containers e.g. FoxyTab).

**Comment by kunaltyagi** [comment]
Is there any roadmap for adding more containers than just the first containers?

**Comment by erosman** [comment]
> Is there any roadmap for adding more containers than just the first containers?

Do you mean more than the first 4?
If there is a popular demand,  it can be done.

However, if Firefox fixes the following, better support can be provided.

> Unfortunately, it is not available as [optional_permissions](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/optional_permissions)
> 
> - [Bug 1386673: Make Contextual Identity extensions be an optional permission](https://bugzilla.mozilla.org/show_bug.cgi?id=1386673)

**Comment by kunaltyagi** [comment]
Last useful update on that feature request was 4 years ago, so hopes aren't that high. Any reason for not using the permission in non-optional mode?

**Comment by erosman** [comment]
> Last useful update on that feature request was 4 years ago, so hopes aren't that high. Any reason for not using the permission in non-optional mode?

As mentioned ...

> - Adding "contextualIdentities" permission automatically enables containers in Firefox which is undesirable to users who don't want to use containers

**Comment by ChristianTacke** [comment]
Can one improve the UI by allowing to change the proxy for "the current container"? That is, the container of the current tab?

foxyproxy seems to somehow know the proxy for the current tab / container, because it can already show the right badge on the empty page (no website) loaded?

Or would this also require this permission?

**Comment by erosman** [comment]
I can think of an alternative approach...

- Remove container option from the Options UI (Incognito setting will remain in Options UI)
- Move the container settings to the toolbar popup

**PRO**
- Container proxy can be set/unset/changed for any current container tab (not only the first 4)

**CON**
- Users must open a container tab first in order to set/unset/change the container proxy

**Comment by ChristianTacke** [comment]
Naive question: Can't one use both? Or is this one or the other (due to different, incompiatible APIs?)?

**Comment by erosman** [comment]
> Naive question: Can't one use both? Or is this one or the other (due to different, incompiatible APIs?)?

Do you mean both on Options page & popup?
It would be duplicating the settings (UI bloatware) and can cause more confusion.

**Comment by ChristianTacke** [comment]
> Do you mean both on Options page & popup?

Yeah, that's what I meant.

>  It would be duplicating the settings (UI bloatware) and can cause more confusion.

Understood!
(I personally like a little UI bloat: The "popup" for more "ad hoc configuration", and the settings page for "dedicated configuration".)

I'd vote for the popup.

The current behaviour on the popup is a little confusing (at least to me): The badge tells me "Using that proxy". While the popup shows "deactivated".

PRO:
* The setting is better connected to the name of the container. Not some random number (you have to know that your work container has number 3.)

**Comment by erosman** [comment]
> Hm... Could anyone explain, what are the downsides of enabling containers?

As I have found out in one of my addons, users who don't use containers, don't want them added/enabled by an extension.

> Make two builds: one with container support, another without.

That is not practical on AMO.

**Comment by d-g** [comment]
> As I have found out in one of my addons, 

FoxyTab?

> users who don't use containers, don't want them added/enabled by an extension

...but donʼt explain, what downsides exactly are?

If anything, webexts, that are quite useful without containers, yet do support them without making a fuss about it, include, for instance:

- https://addons.mozilla.org/en-US/firefox/addon/tree-style-tab/
- https://addons.mozilla.org/en-US/firefox/addon/sidebery/
- https://addons.mozilla.org/en-US/firefox/addon/cookie-quick-manager/

**Comment by erosman** [comment]
> FoxyTab?

Indeed...

> ...but donʼt explain, what downsides exactly are?

Personally, I don't see a problem. However, users don't like it and then complain about it.

**Comment by AdamCalculator** [comment]
up

**Comment by emaba** [comment]
In FoxyProxy, if I set specific proxies for containers, it works as expected only for containers 1 and 2. If I apply it to containers 3 and 4, it doesn't work. I suppose that the issue is with the container ID, but I can't verify it.

**Comment by erosman** [comment]
@emaba Have you changed the default containers?

**Comment by emaba** [comment]
> @emaba Have you changed the default containers?

Yes. Sure.
I for sure changed names and reordered them.

**Comment by erosman** [comment]
> > @emaba Have you changed the default containers?
> 
> Yes. Sure. I for sure changed names and reordered them.

That's is the reason.
FoxyProxy sets it for the ones with original IDs 1-4 (which might not be the first four on your list).

**Comment by emaba** [comment]
> > > @emaba Have you changed the default containers?
> > 
> > 
> > Yes. Sure. I for sure changed names and reordered them.
> 
> That's is the reason. FoxyProxy sets it for the ones with original IDs 1-4 (which might not be the first four on your list).

So, how can I fix it? You should be able to select the ID you want, not only the first four.

**Comment by erosman** [comment]
IDs are set by the browser. For the time being, you can reset the containers to their default and then only change container 5+.

**Comment by emaba** [comment]
> IDs are set by the browser. For the time being, you can reset the containers to their default and then only change container 5+.

And why not allow the user with a simple textbox to choose the ID he wants?

**Comment by erosman** [comment]
> And why not allow the user with a simple textbox to choose the ID he wants?

Do you (and users) know the ID of the container that you want to use?
What are the IDs of your containers?

**Comment by emaba** [comment]
> > And why not allow the user with a simple textbox to choose the ID he wants?
> 
> Do you (and users) know the ID of the container that you want to use? What are the IDs of your containers?

Yes. IDs are in my containers.json file.

**Comment by erosman** [comment]
How many extensions ask their users to go through browser profile folder and open files to be able to perform basic operations?

**Comment by emaba** [comment]
> How many extensions ask their users to go through browser profile folder and open files to be able to perform basic operations?

You are right, but it's really easy to change the behavior.
1) Change the UI to list all the containers:
```
const containers = await browser.contextualIdentities.query({});
containers.forEach(container => {
    const label = document.createElement('label');
    label.textContent = `${container.name} (${container.cookieStoreId})`;
```
2) Enable the selection of a proxy for each container.
3) Save the information exactly as it is currently done.

**Comment by erosman** [comment]
`browser.contextualIdentities` is only available if addon has "contextualIdentities" permission.

**Comment by emaba** [comment]
> `browser.contextualIdentities` is only available if the addon has "contextualIdentities" permission.

Yes. You can ask for permission from the user, too. If the user denies you can revert to the current method.

**Comment by erosman** [comment]
> Yes. You can ask for permission from the user, too. If the user denies you can revert to the current method.

If you read the first post, asking permission i.e. optional permission is not avialable for "contextualIdentities".
It has to be in the manifest and at the time of installation.

**Comment by emaba** [comment]
> > Yes. You can ask for permission from the user, too. If the user denies you can revert to the current method.
> 
> If you read the first post, asking permission i.e. optional permission is not avialable for "contextualIdentities". It has to be in the manifest and at the time of installation.

Thank you. I did not understand that it was not available in optional_permissions.
Why don't we add to the required permissions?

**Comment by erosman** [comment]
> Why don't we add to the required permissions?

Please read the original post. 
Adding the permission, enables containers automatically.
User who don't want to have containers have complained when containers are enabled by an addon without their consent.

**Comment by whoisroot** [comment]
Would it be possible to support more than 4 containers?

I currently have 10 different containers in one profile in Firefox and would very much like to be able to set different proxies between them. More specifically, I have a container (currently container 7) that I really need to have going through a proxy.

The number of containers could easily be parsed from the length of `identities` array in the `containers.json` file of the profile, but I have no idea if this is accessible to the extensions in Firefox (or on Chromium, for that matter). I don't know either if 

In my laziness, I tried exporting the settings, manually editing them to change the container number for which a proxy is set and re-importing it, but wasn't able to make it work.

**Comment by erosman** [comment]
> Would it be possible to support more than 4 containers?

Yes... but where would that end?
Someone can come and ask for 20 containers or more which would create a large & useless user-interface for others with 4 containers.

> The number of containers could easily be parsed from the length of `identities` array in the `containers.json` file of the profile, but I have no idea if this is accessible to the extensions in Firefox (or on Chromium, for that matter).

That file is not accessible to extensions.
Containers are Firefox only feature. They are not available in Chromium.

I have been asking Mozilla engineers to make the container list available to extensions without the additional permission but no change so far.

**Comment by ericjung** [comment]
> Yes... but where would that end?

is it possible to have a text box where users can enter an arbitrary integer?

**Comment by erosman** [comment]
A text box for one is possible, but how many text boxes should there be?
Users can have many containers.

**Comment by ChristianTacke** [comment]
The textbox could be used
- to set the number of container drop downs to show (I only need the first two)
- for typing the container number and then select the proxy next to it. How many of this? Just always show at least one "empty" at the bottom?

No idea, if I my ideas / explanations are any good. Not to mention: No idea if it is possible to implement this at all.

**Comment by erosman** [comment]
For the time being, until Firefox provides a better option, the following is added in v8.10

- Added custom container option 


![image](https://github.com/user-attachments/assets/7ff8fd7b-4405-4019-800b-737d8ad3793c)

**Comment by emaba** [comment]
Thank you!
